staging/src/k8s.io/apiserver/pkg/storage/etcd3/store_strategy.go

```diff
@@ -0,0 +1,217 @@
+type contextKey string
+
+const (
+	// contextKeyCache is the key under which we store a potentially cached object and metadata
+	contextKeyCache = contextKey("cached")
+)
+
+// PotentiallyCachedResponse is a response from etcd which may or may not be stale
+type PotentiallyCachedResponse struct {
+	response *etcdclient.GetResponse
+	cached      bool
+	needsUpdate bool
+}
+
+// preconditionKVClient is a client that knows how to read values from etcd and ensure that they are up-to-date
+// when future transactions are done assuming those values.
+type preconditionKVClient struct {
+	delegate *etcdclient.Client
+
+	guards []TransactionGuard
+}
+
+// Get serves potentially cached data, ensuring that all kvs exposed to the precondition calling this are recorded
+// in modRevision guards so that we know the data used to evaluate the precondition does not change before the
+// transaction is committed.
+func (p *preconditionKVClient) Get(ctx context.Context, key string, opts ...etcdclient.OpOption) (*etcdclient.GetResponse, error) {
+	item := ctx.Value(contextKeyCache)
+	if item == nil {
+		return nil, storage.NewInternalErrorf("found no cached item under key %q in context", contextKeyCache)
+	}
+	cachedResponse, ok := item.(*PotentiallyCachedResponse)
+	if !ok {
+		return nil, storage.NewInternalErrorf("found invalid cached item %T under key %q in context", item, contextKeyCache)
+	}
+	// if we don't have a cached response to return, or we need to update our cached version, fetch from etcd
+	if cachedResponse.response == nil || (cachedResponse.cached && cachedResponse.needsUpdate) {
+		response, err := p.delegate.Get(ctx, key, opts...)
+		if err != nil {
+			return response, err
+		}
+		(*cachedResponse).response = response
+		(*cachedResponse).cached = false
+		(*cachedResponse).needsUpdate = false
+	}
+	for i, kv := range cachedResponse.response.Kvs {
+		p.guards = append(p.guards, &AtModRevision{
+			key:         string(kv.Key),
+			modRevision: kv.ModRevision,
+			consumeUpdatedData: func(response *etcdclient.GetResponse) error {
+				if len(response.Kvs) != 1 {
+					return storage.NewInternalErrorf("invalid gRPC response from etcd: expected 1 response, got %d", len(response.Kvs))
+				}
+				(*cachedResponse).response.Header = response.Header
+				(*cachedResponse).response.Kvs[i] = response.Kvs[0]
+				(*cachedResponse).cached = false
+				return nil
+			},
+		})
+	}
+	return cachedResponse.response, nil
+}
+
+func (p *preconditionKVClient) Guards() []TransactionGuard {
+	return p.guards
+}
+
+var _ PreconditionKVClient = &preconditionKVClient{}
+
+// GuardedTransaction closes over the process of running a transaction against etcd with guards,
+// fetching all the necessary data to evaluate which guards failed, and returning an appropriate
+// error
+type GuardedTransaction struct {
+	ctx    context.Context
+	client *etcdclient.Client
+	guards []TransactionGuard
+
+	// TODO: is this the correct place for these?
```

Comment:
Where else would they go?

Reply:
Seemed like maybe there was a more intelligent way to get the metrics recorded, as we also need such context fields for the `Get()` in preconditions.